client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -349,8 +351,11 @@ public void dataAppended(DataAppended dataAppended) {
             long previousAckLevel = dataAppended.getPreviousEventNumber();
             try {
                 checkAckLevels(ackLevel, previousAckLevel);
-                state.noteSegmentLength(dataAppended.getCurrentSegmentWriteOffset());
-                ackUpTo(ackLevel);
+                // invoke state updation on a different thread and release the EventLoop thread.
+                dataAppendedProcessor.add(() -> CompletableFuture.runAsync(() -> {
+                    state.noteSegmentLength(dataAppended.getCurrentSegmentWriteOffset());
+                    ackUpTo(ackLevel);
```

Comment:
This actually creates a race condition. 
The logic inside checkAckLevels is actually depending on the fact that any previous invocation of this callback has completed `ackUpTo` already because this is what removes the events from the inflight list that it is comparing against.

Reply:
Fixed the code.